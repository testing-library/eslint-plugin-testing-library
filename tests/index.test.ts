import { exec } from 'child_process';
import { existsSync } from 'fs';
import { resolve } from 'path';

import plugin from '../lib';

const generateConfigs = () => exec(`npm run generate:configs`);

const numberOfRules = 26;
const ruleNames = Object.keys(plugin.rules);

// eslint-disable-next-line jest/expect-expect
it('should have a corresponding doc for each rule', () => {
  ruleNames.forEach((rule) => {
    const docPath = resolve(__dirname, '../docs/rules', `${rule}.md`);

    if (!existsSync(docPath)) {
      throw new Error(
        `Could not find documentation file for rule "${rule}" in path "${docPath}"`
      );
    }
  });
});

// eslint-disable-next-line jest/expect-expect
it('should have a corresponding test for each rule', () => {
  ruleNames.forEach((rule) => {
    const testPath = resolve(__dirname, './lib/rules/', `${rule}.test.ts`);

    if (!existsSync(testPath)) {
      throw new Error(
        `Could not find test file for rule "${rule}" in path "${testPath}"`
      );
    }
  });
});

// eslint-disable-next-line jest/expect-expect
it('should have the correct amount of rules', () => {
  const { length } = ruleNames;

  if (length !== numberOfRules) {
    throw new Error(
      `There should be exactly ${numberOfRules} rules, but there are ${length}. If you've added a new rule, please update this number.`
    );
  }
});

it("should have run 'generate:configs' script when changing config rules", async () => {
  generateConfigs();

  const allConfigs = plugin.configs;
  expect(allConfigs).toMatchSnapshot();
});

it('should export configs that refer to actual rules', () => {
  const allConfigs = plugin.configs;

  expect(Object.keys(allConfigs)).toEqual(['dom', 'angular', 'react', 'vue']);
  const allConfigRules = Object.values(allConfigs)
    .map((config) => Object.keys(config.rules))
    .reduce((previousValue, currentValue) => [
      ...previousValue,
      ...currentValue,
    ]);

  allConfigRules.forEach((rule) => {
    const ruleNamePrefix = 'testing-library/';
    const ruleName = rule.slice(ruleNamePrefix.length);

    expect(rule.startsWith(ruleNamePrefix)).toBe(true);
    expect(ruleNames).toContain(ruleName);

    expect(() => require(`../lib/rules/${ruleName}`)).not.toThrow();
  });
});
