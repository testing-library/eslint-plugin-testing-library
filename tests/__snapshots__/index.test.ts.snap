// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have run 'generate:configs' script when changing config rules 1`] = `
Object {
  "angular": Object {
    "plugins": Array [
      "testing-library",
    ],
    "rules": Object {
      "testing-library/await-async-query": "error",
      "testing-library/await-async-utils": "error",
      "testing-library/no-await-sync-query": "error",
      "testing-library/no-container": "error",
      "testing-library/no-debugging-utils": "error",
      "testing-library/no-dom-import": Array [
        "error",
        "angular",
      ],
      "testing-library/no-node-access": "error",
      "testing-library/no-promise-in-fire-event": "error",
      "testing-library/no-render-in-lifecycle": "error",
      "testing-library/no-wait-for-empty-callback": "error",
      "testing-library/no-wait-for-multiple-assertions": "error",
      "testing-library/no-wait-for-side-effects": "error",
      "testing-library/no-wait-for-snapshot": "error",
      "testing-library/prefer-find-by": "error",
      "testing-library/prefer-presence-queries": "error",
      "testing-library/prefer-query-by-disappearance": "error",
      "testing-library/prefer-screen-queries": "error",
      "testing-library/render-result-naming-convention": "error",
    },
  },
  "dom": Object {
    "plugins": Array [
      "testing-library",
    ],
    "rules": Object {
      "testing-library/await-async-query": "error",
      "testing-library/await-async-utils": "error",
      "testing-library/no-await-sync-query": "error",
      "testing-library/no-promise-in-fire-event": "error",
      "testing-library/no-wait-for-empty-callback": "error",
      "testing-library/no-wait-for-multiple-assertions": "error",
      "testing-library/no-wait-for-side-effects": "error",
      "testing-library/no-wait-for-snapshot": "error",
      "testing-library/prefer-find-by": "error",
      "testing-library/prefer-presence-queries": "error",
      "testing-library/prefer-query-by-disappearance": "error",
      "testing-library/prefer-screen-queries": "error",
    },
  },
  "marko": Object {
    "plugins": Array [
      "testing-library",
    ],
    "rules": Object {
      "testing-library/await-async-query": "error",
      "testing-library/await-async-utils": "error",
      "testing-library/await-fire-event": "error",
      "testing-library/no-await-sync-query": "error",
      "testing-library/no-container": "error",
      "testing-library/no-debugging-utils": "error",
      "testing-library/no-dom-import": Array [
        "error",
        "marko",
      ],
      "testing-library/no-node-access": "error",
      "testing-library/no-promise-in-fire-event": "error",
      "testing-library/no-render-in-lifecycle": "error",
      "testing-library/no-unnecessary-act": "error",
      "testing-library/no-wait-for-empty-callback": "error",
      "testing-library/no-wait-for-multiple-assertions": "error",
      "testing-library/no-wait-for-side-effects": "error",
      "testing-library/no-wait-for-snapshot": "error",
      "testing-library/prefer-find-by": "error",
      "testing-library/prefer-presence-queries": "error",
      "testing-library/prefer-query-by-disappearance": "error",
      "testing-library/prefer-screen-queries": "error",
      "testing-library/render-result-naming-convention": "error",
    },
  },
  "react": Object {
    "plugins": Array [
      "testing-library",
    ],
    "rules": Object {
      "testing-library/await-async-query": "error",
      "testing-library/await-async-utils": "error",
      "testing-library/no-await-sync-query": "error",
      "testing-library/no-container": "error",
      "testing-library/no-debugging-utils": "error",
      "testing-library/no-dom-import": Array [
        "error",
        "react",
      ],
      "testing-library/no-node-access": "error",
      "testing-library/no-promise-in-fire-event": "error",
      "testing-library/no-render-in-lifecycle": "error",
      "testing-library/no-unnecessary-act": "error",
      "testing-library/no-wait-for-empty-callback": "error",
      "testing-library/no-wait-for-multiple-assertions": "error",
      "testing-library/no-wait-for-side-effects": "error",
      "testing-library/no-wait-for-snapshot": "error",
      "testing-library/prefer-find-by": "error",
      "testing-library/prefer-presence-queries": "error",
      "testing-library/prefer-query-by-disappearance": "error",
      "testing-library/prefer-screen-queries": "error",
      "testing-library/render-result-naming-convention": "error",
    },
  },
  "vue": Object {
    "plugins": Array [
      "testing-library",
    ],
    "rules": Object {
      "testing-library/await-async-query": "error",
      "testing-library/await-async-utils": "error",
      "testing-library/await-fire-event": "error",
      "testing-library/no-await-sync-query": "error",
      "testing-library/no-container": "error",
      "testing-library/no-debugging-utils": "error",
      "testing-library/no-dom-import": Array [
        "error",
        "vue",
      ],
      "testing-library/no-node-access": "error",
      "testing-library/no-promise-in-fire-event": "error",
      "testing-library/no-render-in-lifecycle": "error",
      "testing-library/no-wait-for-empty-callback": "error",
      "testing-library/no-wait-for-multiple-assertions": "error",
      "testing-library/no-wait-for-side-effects": "error",
      "testing-library/no-wait-for-snapshot": "error",
      "testing-library/prefer-find-by": "error",
      "testing-library/prefer-presence-queries": "error",
      "testing-library/prefer-query-by-disappearance": "error",
      "testing-library/prefer-screen-queries": "error",
      "testing-library/render-result-naming-convention": "error",
    },
  },
}
`;
